rfd/0163-vnet.md

```diff
@@ -0,0 +1,603 @@
+
+* dumper is a web app without `spec.vnet_addr`. "Launch" opens the app in the browser through the
+  `/web/launch` of the proxy service. The user can select "Launch through VNet" from the three dots
+  menu to open `http://dumper.teleport-local` or "Copy VNet address" to copy that address. The
+  address in the card is `https://dumper.teleport-local.dev` as usual.
+* httpbin is a web app with `spec.vnet_addr` set to `httpbin.company.private` and `.company.private`
+  is configured as the suffix of a DNS zone in the `vnet` resource. "Launch" starts VNet if it's not
+  already running and opens the VNet address of the app. The user can select "Launch through Web UI"
+  from the three dots menu. The address in the card is `http://httpbin.company.private`.
+* tcp-postgres is a TCP app without `spec.vnet_addr`. If VNet is not running, the button says
+  "Connect" and it starts VNet, copies the VNet address of the app to the clipboard and shows a
+  notification in the bottom right saying "Copied `tcp-postgres.teleport-local.dev.internal` to the
+  clipboard.". If VNet is running, the button simply says "Copy". The user can select "Set up a
+  legacy connection" which starts the equivalent of `tsh proxy app tcp-postgres` that's available
+  today.
+
+Setting up legacy connections is kept for users who might not be able to use VNet, either because
+they lack root privileges or VNet would conflict with other software running on the computer.
+
+In case of web apps, we don't know if they're intended to be used through VNet or not. The only hint
+we can get is the existence of `spec.vnet_addr` which indicates that the given web app is, for
+example, an HTTP API that users might want to connect to through VNet.
+
+Selecting an app from the search bar performs the same action as clicking on the button next to an
+app.
+
+#### Managing VNet
+
+<img width="871" alt="VNet in Connections" src="assets/0163-vnet-in-connections.png">
+
+VNet has a persistent place in the UI in the connection list in top left. From there, the user can
+quickly stop VNet, similar to how they can disconnect existing database, SSH and Kube connections.
+
+Clicking on the VNet item opens a dedicated VNet tab. From there, the user can see which DNS zones
+are currently being proxied and the status of recent connections.
+
+(TBD, for now see [the comment about capturing errors related to specific
+apps](https://github.com/gravitational/teleport/pull/37834#discussion_r1488830437).)
+
+#### Lifecycle
+
+The first version of VNet integration in Connect behaves similarly to Connect My Computer. After the
+user starts VNet and until they explicitly stop it, VNet starts automatically when launching the
+app. It prompts for a password for root privileges through osascript on each start.
+
+If VNet is running and the user logs in to another cluster, Connect prompts for credentials again
+and configures VNet for that cluster. If this attempt fails, VNet continues to work but the
```

Comment:
@nklaassen Regarding what you said in another comment:

> If DNS is already set up we can just leave it and don't need root again on relogins (but we will need it at least once every time Connect starts to create the TUN and configure the IP routes). This does raise the question of whether we should be removing the DNS configuration every time VNet or connect stops and if there is a good way to do this - we can try to handle the exit signal and clean up.

I agree wrt relogins. It's a fairly common scenario that you log out of a cluster and then log in again. In that case it's essentially equivalent to logging in to a new cluster as we clear all data related to a cluster on logout. So if v1 is to use osascript, we should make it possible to log in again without having to run VNet setup again.

Regarding cleanup, with osascript there just doesn't seem to be a good way to do this. We could make it so that stopping VNet manually prompts for a password and removes our files in `/etc/resolver/`. Cleanup on exit would work only if we left the privileged process running – otherwise we'd have to prompt for root on exit which seems bad. Signaling an exit would be tricky in this case though since we that privileged process runs as a different user. 🤔

With osascript I suppose we could perform minimal cleanup on start by marking all files in `/etc/resolver/` with `# Generated by VNet` and then reading each file and removing those that were generated by us but do not match current clusters.

It's not ideal, but I don't have a better idea. It should be good enough for v1 though.

Reply:
For v1 I think we should start the root process once with osascript and leave it running in the background. It can watch the TSH_HOME dir for any new profiles we log in to and update the host configuration accordingly, there's no need for IPC. It can detect when the original process exits by monitoring the PID or some other means, and then clean up the configuration and exit itself